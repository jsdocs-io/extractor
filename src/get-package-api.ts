import { performance } from "node:perf_hooks";
import { join } from "pathe";
import { Bun } from "./bun.ts";
import { getPackageDeclarations } from "./get-package-declarations.ts";
import { getPackageJson } from "./get-package-json.ts";
import { getPackageOverview } from "./get-package-overview.ts";
import { getPackageTypes } from "./get-package-types.ts";
import { getProject } from "./get-project.ts";
import { tempDir } from "./temp-dir.ts";
import type { ExtractedDeclaration } from "./types.ts";

/** `GetPackageApiOptions` contains the options for calling {@link getPackageApi}. */
export interface GetPackageApiOptions {
	/**
  Package to extract the API from.

  This can be either a package name (e.g., `foo`, `@foo/bar`) or
  any other query that can be passed to `bun add` (e.g., `foo@1.0.0`).

  @see {@link https://bun.sh/docs/cli/add | Bun docs}
  */
	pkg: string;

	/**
  Specific subpath to consider in a package.

  If a package has multiple entrypoints listed in the `exports` map property
  of its `package.json`, use `subpath` to select a specific one by its name
  (e.g., `someFeature`).

  @defaultValue `.` (package root)

  @see {@link https://nodejs.org/api/packages.html#subpath-exports | Node.js docs}
  @see {@link https://github.com/lukeed/resolve.exports | resolve.exports docs}
  */
	subpath?: string;

	/**
  Packages can have deeply nested modules and namespaces.

  Use `maxDepth` to limit the depth of the extraction.
  Declarations nested at levels deeper than this value will be ignored.

  @defaultValue 5
  */
	maxDepth?: number;

	/**
	Bun instance used to install the package.

  @defaultValue a new `Bun` instance
  */
	bun?: Bun;
}

/** `PackageApi` contains the data extracted from a package by calling {@link getPackageApi}. */
export interface PackageApi {
	/** Package name (e.g., `foo`, `@foo/bar`). */
	name: string;

	/** Package version number (e.g., `1.0.0`). */
	version: string;

	/**
  Package subpath selected when extracting the API (e.g., `.`, `someFeature`).

  @see {@link ExtractPackageApiOptions.subpath}
  @see {@link https://nodejs.org/api/packages.html#subpath-exports | Node.js docs}
  */
	subpath: string;

	/**
  Type declarations file, resolved from the selected `subpath`,
  that acts as the entrypoint for the package (e.g., `index.d.ts`).
  */
	types?: string;

	/**
  Package description extracted from the `types` file if a
  JSDoc comment with the `@packageDocumentation` tag is found.
  */
	overview?: string;

	/** Declarations exported (or re-exported) by the package. */
	declarations: ExtractedDeclaration[];

	/**
  All packages resolved and installed when installing the package (included).

  @example
  ```ts
  // Installing `foo` brings in also `bar` and `baz` as dependencies.
  ["foo@1.0.0", "bar@2.0.0", "baz@3.0.0"]
  ```
  */
	dependencies: string[];

	/** Timestamp of when the package was analyzed. */
	analyzedAt: string;

	/** Package analysis duration in milliseconds. */
	analyzedIn: number;
}

/**
`getPackageApi` extracts the API from a package.

If the extraction succeeds, `getPackageApi` returns a {@link PackageApi} object.
If the extraction fails, `getPackageApi` throws an error.

Warning: The extraction process is slow and blocks the main thread, using workers is recommended.

@example
```ts
const packageApi = await getPackageApi({
  pkg: "foo",    // Extract API from npm package `foo` [Required]
  subpath: ".",  // Select subpath `.` (root subpath) [Optional]
  maxDepth: 5,   // Maximum depth for analyzing nested namespaces [Optional]
  bun: new Bun() // Bun package manager instance [Optional]
});
console.log(JSON.stringify(packageApi, null, 2));
```

@param options - {@link GetPackageApiOptions}

@returns A {@link PackageApi} object
*/
export async function getPackageApi({
	pkg,
	subpath = ".",
	maxDepth = 5,
	bun = new Bun(),
}: GetPackageApiOptions): Promise<PackageApi> {
	// Normalize options.
	pkg = pkg.trim();
	subpath = subpath.trim() || ".";
	maxDepth = Math.max(1, Math.round(maxDepth));

	// Start performance timer.
	const start = performance.now();

	// Create a temporary directory where to install the package.
	await using dir = await tempDir();
	const cwd = dir.path;

	// Install the package and its direct and third-party dependencies.
	const dependencies = await bun.add(pkg, cwd);

	// Read the package's `package.json`.
	const pkgName = await getInstalledPackageName(cwd);
	const pkgDir = join(cwd, "node_modules", pkgName);
	const pkgJson = await getPackageJson(pkgDir);
	const { name, version } = pkgJson;

	// Find the package's types entry point file, if any.
	const types = getPackageTypes({ pkgJson, subpath });
	if (!types) {
		return {
			name,
			version,
			subpath,
			types: undefined,
			overview: undefined,
			declarations: [],
			dependencies,
			analyzedAt: analyzedAt(),
			analyzedIn: analyzedIn(start),
		};
	}

	// Create TypeScript project.
	const pkgTypes = join(pkgDir, types);
	const { project, indexFile } = getProject({ indexFilePath: pkgTypes, typeRoots: cwd });

	// Get overview.
	const overview = getPackageOverview(indexFile);

	// Extract the declarations exported by the package.
	const declarations = await getPackageDeclarations({ pkgName, project, indexFile, maxDepth });

	// Return the data extracted from the package.
	return {
		name,
		version,
		subpath,
		types,
		overview,
		declarations,
		dependencies,
		analyzedAt: analyzedAt(),
		analyzedIn: analyzedIn(start),
	};
}

async function getInstalledPackageName(cwd: string): Promise<string> {
	// Since `pkg` can contain any argument accepted by Bun's `add` command
	// (e.g., URLs), get the package name from the only dependency listed in
	// the `package.json` file created by Bun in the `cwd` on install.
	const pkgJson = await getPackageJson(cwd);
	return Object.keys(pkgJson.dependencies!).at(0)!;
}

function analyzedAt(): string {
	return new Date().toISOString();
}

function analyzedIn(start: number): number {
	return Math.round(performance.now() - start);
}
