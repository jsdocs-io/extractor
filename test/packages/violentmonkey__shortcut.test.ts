import { expect, test } from "vitest";
import { getPackageApi } from "../../src/get-package-api.ts";

test("@violentmonkey/shortcut@1.4.1", async () => {
	expect(
		await getPackageApi({
			pkg: "@violentmonkey/shortcut@1.4.1",
		}),
	).toMatchSnapshot({
		analyzedAt: expect.any(String),
		analyzedIn: expect.any(Number),
		dependencies: expect.any(Array),
	});
});
