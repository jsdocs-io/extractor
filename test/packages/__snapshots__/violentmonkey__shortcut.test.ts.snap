// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@violentmonkey/shortcut@1.4.1 1`] = `
{
  "analyzedAt": Any<String>,
  "analyzedIn": Any<Number>,
  "declarations": [
    {
      "constructors": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.constructor",
          "kind": "class-constructor",
          "line": 16,
          "name": "constructor",
          "signature": "constructor(options?: Partial<IShortcutServiceOptions>);",
        },
      ],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/index.d.ts",
      "id": "+class.KeyboardService",
      "kind": "class",
      "line": 6,
      "methods": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+method.disable",
          "kind": "class-method",
          "line": 25,
          "name": "disable",
          "signature": "disable: () => void;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+method.enable",
          "kind": "class-method",
          "line": 24,
          "name": "enable",
          "signature": "enable: () => void;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+method.handleKeyOnce",
          "kind": "class-method",
          "line": 28,
          "name": "handleKeyOnce",
          "signature": "handleKeyOnce: (keyExps: string[], fromRoot: boolean) => boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+method.register",
          "kind": "class-method",
          "line": 26,
          "name": "register",
          "signature": "register: (
  key: string,
  callback: () => void,
  options?: Partial<IShortcutOptions>,
) => () => void;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+method.setContext",
          "kind": "class-method",
          "line": 27,
          "name": "setContext",
          "signature": "setContext: (key: string, value: unknown) => void;",
        },
      ],
      "name": "KeyboardService",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+property.defaultOptions",
          "kind": "class-property",
          "line": 14,
          "name": "defaultOptions",
          "signature": "static defaultOptions: IShortcutServiceOptions;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+property.handleKey",
          "kind": "class-property",
          "line": 29,
          "name": "handleKey",
          "signature": "handleKey: (e: KeyboardEvent) => void;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+property.options",
          "kind": "class-property",
          "line": 15,
          "name": "options",
          "signature": "options: IShortcutServiceOptions;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/index.d.ts",
          "id": "+class.KeyboardService.+property.sequence",
          "kind": "class-property",
          "line": 12,
          "name": "sequence",
          "signature": "sequence: Subject<string[]>;",
        },
      ],
      "signature": "class KeyboardService {}",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/util.d.ts",
      "id": "+function.buildKey",
      "kind": "function",
      "line": 2,
      "name": "buildKey",
      "signature": "buildKey: (key: IShortcutKey) => string;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/index.d.ts",
      "id": "+function.disable",
      "kind": "function",
      "line": 34,
      "name": "disable",
      "signature": "disable: () => void;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/index.d.ts",
      "id": "+function.enable",
      "kind": "function",
      "line": 33,
      "name": "enable",
      "signature": "enable: () => void;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/index.d.ts",
      "id": "+function.getService",
      "kind": "function",
      "line": 31,
      "name": "getService",
      "signature": "getService: () => KeyboardService;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/index.d.ts",
      "id": "+function.handleKey",
      "kind": "function",
      "line": 35,
      "name": "handleKey",
      "signature": "handleKey: (e: KeyboardEvent) => void;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/util.d.ts",
      "id": "+function.normalizeSequence",
      "kind": "function",
      "line": 4,
      "name": "normalizeSequence",
      "signature": "normalizeSequence: (
  input: string | string[],
  caseSensitive: boolean,
) => IShortcutKey[];",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/util.d.ts",
      "id": "+function.parseCondition",
      "kind": "function",
      "line": 5,
      "name": "parseCondition",
      "signature": "parseCondition: (condition: string) => IShortcutCondition[];",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/util.d.ts",
      "id": "+function.parseKey",
      "kind": "function",
      "line": 3,
      "name": "parseKey",
      "signature": "parseKey: (shortcut: string, caseSensitive: boolean) => IShortcutKey;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/index.d.ts",
      "id": "+function.register",
      "kind": "function",
      "line": 32,
      "name": "register",
      "signature": "register: (
  key: string,
  callback: () => void,
  options?: Partial<IShortcutOptions> | undefined,
) => () => void;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/util.d.ts",
      "id": "+function.reprKey",
      "kind": "function",
      "line": 6,
      "name": "reprKey",
      "signature": "reprKey: (key: IShortcutKey) => string;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/util.d.ts",
      "id": "+function.reprShortcut",
      "kind": "function",
      "line": 7,
      "name": "reprShortcut",
      "signature": "reprShortcut: (input: string | string[], caseSensitive?: boolean) => string;",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IKeyNode",
      "indexSignatures": [],
      "kind": "interface",
      "line": 18,
      "methods": [],
      "name": "IKeyNode",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IKeyNode.+property.children",
          "kind": "interface-property",
          "line": 20,
          "name": "children",
          "signature": "children: Map<string, IKeyNode>;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IKeyNode.+property.shortcuts",
          "kind": "interface-property",
          "line": 19,
          "name": "shortcuts",
          "signature": "shortcuts: Set<IShortcut>;",
        },
      ],
      "setAccessors": [],
      "signature": "interface IKeyNode {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcut",
      "indexSignatures": [],
      "kind": "interface",
      "line": 11,
      "methods": [],
      "name": "IShortcut",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcut.+property.callback",
          "kind": "interface-property",
          "line": 14,
          "name": "callback",
          "signature": "callback: () => void;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcut.+property.caseSensitive",
          "kind": "interface-property",
          "line": 16,
          "name": "caseSensitive",
          "signature": "caseSensitive: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcut.+property.condition",
          "kind": "interface-property",
          "line": 13,
          "name": "condition",
          "signature": "condition?: string;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcut.+property.enabled",
          "kind": "interface-property",
          "line": 15,
          "name": "enabled",
          "signature": "enabled: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcut.+property.sequence",
          "kind": "interface-property",
          "line": 12,
          "name": "sequence",
          "signature": "sequence: string[];",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcut {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcutCondition",
      "indexSignatures": [],
      "kind": "interface",
      "line": 7,
      "methods": [],
      "name": "IShortcutCondition",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutCondition.+property.field",
          "kind": "interface-property",
          "line": 8,
          "name": "field",
          "signature": "field: string;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutCondition.+property.not",
          "kind": "interface-property",
          "line": 9,
          "name": "not",
          "signature": "not: boolean;",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcutCondition {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcutConditionCache",
      "indexSignatures": [],
      "kind": "interface",
      "line": 22,
      "methods": [],
      "name": "IShortcutConditionCache",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutConditionCache.+property.count",
          "kind": "interface-property",
          "line": 23,
          "name": "count",
          "signature": "count: number;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutConditionCache.+property.result",
          "kind": "interface-property",
          "line": 25,
          "name": "result",
          "signature": "result: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutConditionCache.+property.value",
          "kind": "interface-property",
          "line": 24,
          "name": "value",
          "signature": "value: IShortcutCondition[];",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcutConditionCache {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcutKey",
      "indexSignatures": [],
      "kind": "interface",
      "line": 35,
      "methods": [],
      "name": "IShortcutKey",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutKey.+property.base",
          "kind": "interface-property",
          "line": 36,
          "name": "base",
          "signature": "base: string;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutKey.+property.caseSensitive",
          "kind": "interface-property",
          "line": 38,
          "name": "caseSensitive",
          "signature": "caseSensitive: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutKey.+property.modifierState",
          "kind": "interface-property",
          "line": 37,
          "name": "modifierState",
          "signature": "modifierState: IShortcutModifiers;",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcutKey {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcutModifiers",
      "indexSignatures": [],
      "kind": "interface",
      "line": 1,
      "methods": [],
      "name": "IShortcutModifiers",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutModifiers.+property.a",
          "kind": "interface-property",
          "line": 4,
          "name": "a",
          "signature": "a?: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutModifiers.+property.c",
          "kind": "interface-property",
          "line": 2,
          "name": "c",
          "signature": "c?: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutModifiers.+property.m",
          "kind": "interface-property",
          "line": 5,
          "name": "m",
          "signature": "m?: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutModifiers.+property.s",
          "kind": "interface-property",
          "line": 3,
          "name": "s",
          "signature": "s?: boolean;",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcutModifiers {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcutOptions",
      "indexSignatures": [],
      "kind": "interface",
      "line": 27,
      "methods": [],
      "name": "IShortcutOptions",
      "properties": [
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutOptions.+property.caseSensitive",
          "kind": "interface-property",
          "line": 29,
          "name": "caseSensitive",
          "signature": "caseSensitive: boolean;",
        },
        {
          "docs": [],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutOptions.+property.condition",
          "kind": "interface-property",
          "line": 28,
          "name": "condition",
          "signature": "condition?: string;",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcutOptions {}",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
      "getAccessors": [],
      "id": "+interface.IShortcutServiceOptions",
      "indexSignatures": [],
      "kind": "interface",
      "line": 31,
      "methods": [],
      "name": "IShortcutServiceOptions",
      "properties": [
        {
          "docs": [
            "/** Max timeout between two keys within a sequence. */",
          ],
          "file": "/@violentmonkey/shortcut/types/types/index.d.ts",
          "id": "+interface.IShortcutServiceOptions.+property.sequenceTimeout",
          "kind": "interface-property",
          "line": 33,
          "name": "sequenceTimeout",
          "signature": "sequenceTimeout: number;",
        },
      ],
      "setAccessors": [],
      "signature": "interface IShortcutServiceOptions {}",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/constants.d.ts",
      "id": "+type.IModifier",
      "kind": "type",
      "line": 2,
      "name": "IModifier",
      "signature": "type IModifier = (typeof modifierList)[number];",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/constants.d.ts",
      "id": "+variable.aliases",
      "kind": "variable",
      "line": 11,
      "name": "aliases",
      "signature": "const aliases: Record<string, string>;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/constants.d.ts",
      "id": "+variable.modifierAliases",
      "kind": "variable",
      "line": 4,
      "name": "modifierAliases",
      "signature": "const modifierAliases: Record<string, "m" | "c" | "s" | "a">;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/constants.d.ts",
      "id": "+variable.modifierList",
      "kind": "variable",
      "line": 1,
      "name": "modifierList",
      "signature": "const modifierList: readonly ["m", "c", "s", "a"];",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/constants.d.ts",
      "id": "+variable.modifiers",
      "kind": "variable",
      "line": 3,
      "name": "modifiers",
      "signature": "const modifiers: Record<string, "m" | "c" | "s" | "a">;",
    },
    {
      "docs": [],
      "file": "/@violentmonkey/shortcut/types/constants.d.ts",
      "id": "+variable.modifierSymbols",
      "kind": "variable",
      "line": 5,
      "name": "modifierSymbols",
      "signature": "const modifierSymbols: { c: string; s: string; a: string; m: string };",
    },
  ],
  "dependencies": Any<Array>,
  "name": "@violentmonkey/shortcut",
  "overview": undefined,
  "subpath": ".",
  "types": "types/index.d.ts",
  "version": "1.4.1",
}
`;
